### PR TITLE
Lambda Scheduled Provisioned Concurrency - Closes #24

### LINKED ISSUE #24: Ability to Provision a lambda between 8 AM to 8 PM and turn it unprovisioned during Non Business hours
Hi, 

Could you please add Lambda Provision Scheduling, this will help provision lambdas during the normal business hours and turn it off during non business hours.

Please find the AWS Blog for reference: https://aws.amazon.com/blogs/compute/scheduling-aws-lambda-provisioned-concurrency-for-recurring-peak-usage/
Automate the scheduling of Lambda Provisioned Concurrency with 
```
      ScheduledActions:                                  
        - ScalableTargetAction:
            MinCapacity: 250
          Schedule: 'cron(45 11 * * ? *)'
          ScheduledActionName: scale-out

```
Thank you,
Naveen

### PR DESCRIPTION
**Pull request by cookieMr**

### CODE REVIEW

**Comment by claydanford** [comment]
@nmrad-91 Let's hold on this until I can give a closer look at it. My other pr still needs to be merged.

**Comment by cookieMr** [comment]
yeap, i planed to wait for your changes to go 1st

**Comment by claydanford** [comment]
How does the scheduled action interact with the target tracking policies?

File: src/aws/target.ts
Comment (claydanford) [inline src/aws/target.ts:19]: Can you remove these console logs?

**Comment by nmrad-91** [comment]
@cookieMr  and @claydanford I will hold off on this one until I hear from you two. Sorry about the conflict. I should have merged this yesterday.

File: src/aws/target.ts
Comment (claydanford) [inline src/aws/target.ts:55]: Instead of creating this object, can we use a ternary, like in the custom metric defaults, in my other pr?

```typescript
ScheduledActions: this.data.scheduledActions
  ? this.getSchedulesActions()
  : undefined,
```

**Comment by cookieMr** [comment]
@nmrad-91 yes, please hold on, i'll reach out to you when this will be ready
@claydanford i'd like you to approve this before i merge (but 1st let's talk)

@claydanford 
> How does the scheduled action interact with the target tracking policies?

[Target tracking scaling policies for Application Auto Scaling](https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-target-tracking.html); Chapter *Supporting application availability during high utilization periods*:

> For workloads that are cyclical in nature, you also have the option to automate capacity changes on a schedule using scheduled scaling. For each scheduled action, a new minimum capacity value and a new maximum capacity value can be defined. These values form the boundaries of the scaling policy.
>
> **The combination of scheduled scaling and target tracking scaling can help reduce the impact of a sharp increase in utilization levels, when capacity is needed immediately.**

Looks like target tracking policy will smooth the slace out/in action defined by scheduled actions.

File: src/aws/target.ts
Comment (cookieMr) [reply src/aws/target.ts]: yeap, np
done

**Comment by claydanford** [comment]
@cookieMr what configuration are you using to test? I'm getting validation errors.

**Comment by cookieMr** [comment]
I use something like this and i am able to generate CloudFormation code.

```yaml
functions:
  dummy:
    handler: src/DummyHandler.dummy
    provisionedConcurrency: 1
    concurrencyAutoscaling:
      enabled: true
      min: 1
      max: 10
      scheduledActions:
        - name: OpenOfficeTime
          schedule: cron(* 30 8 * 1-6 *)
          action:
            minimum: 10
        - name: CloseOfficeTime
          schedule: cron(* 30 17 * 1-6 *)
          action:
            minimum: 5
```
```json
    "DummyAutoScalingTargetDevUswest2": {
      "DependsOn": [
        "CummyProvConcLambdaAlias"
      ],
      "Properties": {
        "MaxCapacity": 10,
        "MinCapacity": 1,
        "ResourceId": "function:dummy:provisioned",
        "ScalableDimension": "lambda:function:ProvisionedConcurrency",
        "ServiceNamespace": "lambda",
        "RoleARN": {
          "Fn::Sub": "arn:aws:iam::${AWS::AccountId}:role/aws-service-role/lambda.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_LambdaConcurrency"
        },
        "ScheduledActions": [
          {
            "ScalableTargetAction": {
              "MaxCapacity": 10
            },
            "ScheduledActionName": "OpenOfficeTime",
            "Schedule": "cron(* 30 8 * 1-6 *)"
          },
          {
            "ScalableTargetAction": {
              "MaxCapacity": 5
            },
            "ScheduledActionName": "CloseOfficeTime",
            "Schedule": "cron(* 30 17 * 1-6 *)"
          }
        ]
      },
      "Type": "AWS::ApplicationAutoScaling::ScalableTarget"
    },
```

and when i do the AWS CF command to verify SLS CF template:
```bash
aws cloudformation validate-template \
    --region=us-west-2 \
    --template-body file://cloudformation-template-update-stack.json
```
i get no errors/issues.

But i didn't deployed this into AWS (yet).

**Comment by claydanford** [comment]
I'm getting all sorts of validation exceptions with the cron or schedule. Happy for this to go through, but that should be tested first. Good work by the way.

**Comment by cookieMr** [comment]
> I'm getting all sorts of validation exceptions with the cron or schedule.

Can you share these exceptions?

> Happy for this to go through, but that should be tested first.

I am not merging without tests. I still need some time to test.

**Comment by claydanford** [comment]
Error:
- `Invalid schedule expression. Details: Invalid CRON expression: error at position 0, details: Specifying both a day-of-week AND a day-of-month parameter is not supported.`
- `Invalid schedule frequency. Run for '1970-01-01T17:30:01.000Z' and '1970-01-01T17:30:00.000Z' are less than 60 seconds apart`

I think we just need some good valid examples, my time is limited right now.

**Comment by cookieMr** [comment]
I see the issue. It's the `cron` syntax. [According to AWS doc](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html#CronExpressions) (section `Restrictions`) `cron` expression cannot have asterisk everywhere.

> You can't specify the Day-of-month and Day-of-week fields in the same cron expression. If you specify a value (or a *) in one of the fields, you must use a ? (question mark) in the other.

Actually I was surprised to see any issue with CloudFormation template since I took it from [AWS itself](https://aws.amazon.com/blogs/compute/scheduling-aws-lambda-provisioned-concurrency-for-recurring-peak-usage/).

Also, i wonder if i should extract documentation (readme) about `cron` and rate (etc.) to a separate markdown file (and refer to it in the readme).

**Comment by claydanford** [comment]
I would link to aws documentation in case it changes.

**Comment by cookieMr** [comment]
@claydanford 
> I would link to aws documentation in case it changes.

i already did this, the 2nd bullet ***before*** `Known Issues/Limitations` section.

**Comment by jsyoo** [comment]
@claydanford is this pr still pending?

**Comment by cookieMr** [comment]
@jsyoo yes, this is still pending, i have little time to test it on AWS and fix 1 minor issue i found (min/max for scheduled actions), i still plan to finish it

**Comment by cookieMr** [comment]
@claydanford this is done. Ready. I was able to successfully do a deployment with all 4 scheduled actions from a new readme file.

**Comment by claydanford** [comment]
Hi @cookieMr @nmrad-91, will review and report back.

**Comment by claydanford** [comment]
Using this config, I am not able to recreate with your latest changes in the master branch

```YAML
functions:
  hello:
    handler: handler.hello
    provisionedConcurrency: 1
    concurrencyAutoscaling:
      enabled: true
      minimum: 1
      maximum: 10
      scheduledActions:
        - name: OpenOfficeTime
          startTime: "2025-01-01T00:00:00.000Z"
          endTime: "2025-01-01T23:59:59.999Z"
          timezone: "America/Chicago"
          schedule: "cron(30 8 ? * 1-6 *)"
          action:
            maximum: 10
            minimum: 1
        - name: CloseOfficeTime
          startTime: "2025-01-01T00:00:00.000Z"
          endTime: "2025-01-01T23:59:59.999Z"
          timezone: "America/Chicago"
          schedule: "cron(30 17 ? * 1-6 *)"
          action:
            maximum: 10
            minimum: 1
```

**Comment by claydanford** [comment]
Okay this looks good. I was able to get it to deploy. Since the console doesn't show these schedules I cannot verify it works as expected, but it does deploy.

**Review by claydanford — APPROVED** [review]

**Comment by cookieMr** [comment]
@claydanford 
when i was testing this i saw Lambda's PC was increased by this scheduled action
but i couldn't see it to decrease, i presume this is due to the issue you already wrote about in #36 

P.S.
Don't use the same values for increase/decrease scheduled actions as this might seems like actions are not happening

**Comment by cookieMr** [comment]
@claydanford in your opinion, is this PR ready to merge and rolled out to the world?

i did all test i can thing of in AWS and it works (see my previous comment here)

**Comment by claydanford** [comment]
@cookieMr sounds good merge and I will publish!

**Comment by cookieMr** [comment]
Sorry, i don't have that power (of merge)
but i have a way to contact @nmrad-91 :)

**Comment by claydanford** [comment]
@cookieMr v1.6.0 published to NPM!

**Comment by cookieMr** [comment]
now i need to support/maintain it

**Pull request merged**
